packages/jupyter-ai/jupyter_ai/personas/base_persona.py

```diff
@@ -0,0 +1,251 @@
+        returned by `self.as_user()`. Provided by `BasePersona`.
+
+        NOTE/TODO: This currently just returns the value set in `self.defaults`.
+        This is set here because we may require this field to be configurable
+        for all personas in the future.
+        """
+        return self.defaults.avatar_path
+
+    @property
+    def system_prompt(self) -> str:
+        """
+        Returns the system prompt used by this persona. Provided by `BasePersona`.
+
+        NOTE/TODO: This currently just returns the value set in `self.defaults`.
+        This is set here because we may require this field to be configurable
+        for all personas in the future.
+        """
+        return self.defaults.system_prompt
+
+    def as_user(self) -> User:
+        """
+        Returns the `jupyterlab_chat.models:User` model that represents this
+        persona in the chat. This model also includes all attributes from
+        `jupyter_server.auth:JupyterUser`, the user model returned by the
+        `IdentityProvider` in Jupyter Server.
+
+        This method is provided by `BasePersona`.
+        """
+        return User(
+            username=self.id,
+            name=self.name,
+            display_name=self.name,
+            avatar_url=self.avatar_path,
+        )
+
+    def as_user_dict(self) -> Dict[str, Any]:
+        """
+        Returns `self.as_user()` as a Python dictionary. This method is provided
+        by `BasePersona`.
+        """
+        user = self.as_user()
+        return asdict(user)
+
+    async def forward_reply_stream(self, reply_stream: "AsyncIterator"):
+        """
+        Forwards an async iterator, dubbed the 'reply stream', to a new message
+        by this persona in the YChat.
+
+        - Creates a new message upon receiving the first chunk from the reply
+        stream, then continuously updates it until the stream is closed.
+
+        - Automatically manages its awareness state to show writing status.
+        """
+        stream_id: Optional[str] = None
+
+        try:
+            self.awareness.set_local_state_field("isWriting", True)
+            async for chunk in reply_stream:
+                if not stream_id:
+                    stream_id = self.ychat.add_message(
+                        NewMessage(body="", sender=self.id)
+                    )
+
+                assert stream_id
+                self.ychat.update_message(
+                    Message(
+                        id=stream_id,
+                        body=chunk,
+                        time=time(),
+                        sender=self.id,
+                        raw_time=False,
+                    ),
+                    append=True,
+                )
+        except Exception as e:
+            self.log.error(
+                f"Persona '{self.name}' encountered an exception printed below when attempting to stream output."
+            )
+            self.log.exception(e)
```

Comment:
Why 2 separate statements? Wouldn't `log.exception` do both?
```python
self.log.exception(
  f"Persona '{self.name}' encountered an exception printed below when attempting to stream output."
)
```

Reply:
You're right. I didn't know that `self.log.exception()` could get the exception details from the surrounding context. Fixed in latest commit.